Enhance determination of uniform writability in Vulkan RD

RandomShaper opened the pull request:
~The main goal of this PR is to add a sanity check.~ **UPDATE 1:** Now this PR adds two improvements to the whole descriptor writability area, so it's now not only a asanity check. Quoting the commit where those happen:
> - Check block decoration in addition to type decoration to be sure to find `readonly` decorators
> - Verify uniforms have same writability across all shader stages in Vulkan RD

Since it's ~theoretically~ (**UPDATE 1:** Now it's also practically) a good idea to re-create the shader cache with that check added, I'm taking the chance to add a little feature that would allow developers to force the regeneration of the cache, as I discussed with @Calinou.

~**UPDATE 1:** Tagging as _Needs work_ since now I have to fix a case of the same-writability rule in descriptor sets not being obeyed. Will add and untag.~

**UPDATE 2:** Pending commit already added: SFDGI direct lighting was violating the same-writability rule. Formerly it went unnoticed because some cases of read-only descriptors were not being detected, something which this PR fixes.

**UPDATE 3:** Changed the major.minor shader cache version scheme I added in some commit by one based in Godot version and commit hash.

Fixes #63056.
Fixes #62525.

clayjohn reviewed (commented):
Looks fine to me, but will also need a sanity check from @reduz

RandomShaper committed:
Enhance determination of uniform writability in Vulkan RD

- Check block decoration in addition to type decoration to be sure to find `readonly` decorators
- Verify uniforms have same writability across all shader stages in Vulkan RD

RandomShaper committed:
Make SDFGI direct light shader follow the same-writability rule

Calinou commented:
> Since it's theoretically (UPDATE 1: Now it's also practically) a good idea to re-create the shader cache with that check added, I'm taking the chance to add a little feature that would allow developers to force the regeneration of the cache, as I discussed with @Calinou.

I opened a dedicated issue to track this: https://github.com/godotengine/godot/issues/63056

RandomShaper committed:
Improve versioning of shader binary data files

- Include Godot version and commit hash in shader cache key
- Reject files when format doesn't match, even if it's lower, since we don't have backwards compatibility here
